Add switch to change from biblatex to bibtex and vice versa (fi…

Ka0o0 opened the pull request:
Hi,
this PR adds an additional menu item to switch from biblatex to bibtex and vice versa. As discussed in the issue, I first wanted to add another combobox to the library properties pane but I noticed that in the save actions was a button which depended on the selected.
To keep things simple for 5.0 I rather adopted the way things used to work in Jabref 4.x by adding another menu item.

Any suggestions on this solution?

All the best

----

- [x] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [x] Screenshots added in PR description (for bigger UI changes)
- [x] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

**Updated Sreen:**

<img width="1033" alt="Screen Shot 2019-11-06 at 18 21 11" src="https://user-images.githubusercontent.com/4035371/68322945-acbed900-00c4-11ea-92cb-93bbc45803d5.png">

Siedlerchr commented on src/main/java/org/jabref/gui/editmode/EditModeAction.java:
```diff
@@ -0,0 +1,47 @@
+
+import static org.jabref.gui.actions.ActionHelper.needsDatabase;
+
+public class EditModeAction extends SimpleCommand {
```
I would suggest renaming it to EditBibDatabaseMode to have it a bit more clear which mode is meant.
Otherwise lgtm! Thanks  for your ongoing contribution!

Ka0o0 commented:
@Siedlerchr and @tobiasdiez I've now added another combo box to the library properties pane.
I've also changed the name of the button to "Reset to recommended".

However, now there is reset and Reset to recommended. Isn't it a bit confusing for the user?

<img width="1033" alt="Screen Shot 2019-11-06 at 18 21 11" src="https://user-images.githubusercontent.com/4035371/68322945-acbed900-00c4-11ea-92cb-93bbc45803d5.png">

Siedlerchr commented:
I would suggest renaming "Reset" to "clear" to avoid confusion. 

I personally would have expected a radio box for switching between biblatex and bibtex but a combox is fine, too.

koppor commented:
I would have expected a toggle switch - or a slider with two settings only.

https://controlsfx.bitbucket.io/org/controlsfx/control/ToggleSwitch.html

Keeping as is (drop down) is very fine for me.

+1 for renaming "Reset" to "Clear"!

tobiasdiez commented:
I'm also in favor of renaming `Reset` to `Clear`. But then also the behavior needs to be changed to really clear the list. (The current behavior is strange anyway...why do we have to different sets of recommended/default actions.)

Ka0o0 commented:
I would then suggest to also change the order of the buttons to keep clearing actions (remove selected and remove all) together. How about having

Reset to recommended - Remove selected - Remove all

tobiasdiez commented on src/main/java/org/jabref/gui/cleanup/FieldFormatterCleanupsPanel.java:
```diff
@@ -115,7 +115,7 @@ private void buildLayout() {
 
-        recommendButton = new Button(Localization.lang("Recommended for %0", databaseContext.getMode().getFormattedName()));
+        recommendButton = new Button(Localization.lang("Reset to recommended", databaseContext.getMode().getFormattedName()));
         boolean isBiblatex = databaseContext.isBiblatexMode();
```
Please move this line in the `setOnAction` block so that changes to the mode are reflected in the buttons behavior.

Ka0o0 replied:
Sorry, I don't quite get it. Do you mean because of the change of name? This is why I renamed the button so there is no reference to the used mode. Or do you mean something else?

tobiasdiez replied:
Sorry for being imprecise here. In fact, I don't know what I meant either 😸

tobiasdiez commented on src/main/java/org/jabref/gui/libraryproperties/LibraryPropertiesDialogView.java:
```diff
@@ -124,6 +129,13 @@ private void storeSettings() {
 
+        BibDatabaseMode newMode = BibDatabaseMode.parse(viewModel.selectedDatabaseModeProperty().getValue());
+        BibDatabaseMode oldMode = metaData.getMode().orElse(null);
+        if (newMode != null && oldMode != newMode) {
```
can you please move this code to `metadata.setMode`. Setting the mode to the same mode should not trigger any actions.

tobiasdiez commented on src/main/java/org/jabref/gui/libraryproperties/LibraryPropertiesDialogView.java:
```diff
@@ -33,6 +34,7 @@
 
     @FXML private VBox contentVbox;
     @FXML private ComboBox<Charset> encoding;
+    @FXML private ComboBox<String> databaseMode;
```
If you use `ComboBox<BibDatabaseMode>` here (and propagate the change), then you don't have to constantly convert between formatted values and parsed modes.

Ka0o0 replied:
If I do so, the names are not formatted anymore. Also, the view would have knowledge about the model.
I normally always try to avoid that by converting the model to non model datatypes. So I would suggest to leave it this way, what do you think?

tobiasdiez replied:
You can get the proper display quite easily using similar code as the one below:
https://github.com/JabRef/jabref/blob/a76b8f3a3814b54cadcaad1826add136b33f2c98/src/main/java/org/jabref/gui/cleanup/FieldFormatterCleanupsPanel.java#L174-L177

You are right, one should usually not use model classes directly in the ui. So here the "correct" solution would be to wrap the `DatabaseMode` in a `DatabaseModeViewModel` which exposes the `getFormattedName` method. I feel like this is however overkill for this particular situation.

Ka0o0 commented:
Before this PR is getting merged, I want to discuss the following:
Should we consider updating the save actions if the user changes the database mode and he still has the default/recommended save actions set?

calixtus commented:
Hi, Ka0o0.
About the buttons: since I saw that iconbuttons were used in some places, I implemented them also in the preferences dialog. Maybe this an idea here too? You could also implement the "remove selected" button in an action column on the right. Just an idea, so the ui would be a bit more consistent.

Ka0o0 commented:
Hi @calixtus,
yes but I think, we should try to keep this PR small and discuss this separately. Maybe you can create another suggestion issue then?

tobiasdiez commented:
> Should we consider updating the save actions if the user changes the database mode and he still has the default/recommended save actions set?

I think this would be a neat addition!

koppor commented:
@Ka0o0 +1 for changing the default save actions. This would IMHO be very intuitive. - Only pro users change the save actions.

Refs https://github.com/JabRef/jabref/pull/2051

Refs https://github.com/JabRef/jabref/pull/2601 - follow-up-issue: https://github.com/JabRef/jabref/issues/3644

Ka0o0 commented:
Okay, great. I will create an issue for this then.

koppor commented:
@Ka0o0 I currently see following checklist at the PR description:

![grafik](https://user-images.githubusercontent.com/1366654/68532352-767b9680-031c-11ea-94fe-f204c0f32fea.png)

Could you update this regarding what is missing? - Background: We go through the PRs and check for the status. It would be very helpful if the checklist was modified w.r.t. the status. - No over-engineering here. If you just have to work again, you can prefix the PR title with [`[WIP] `](https://stackoverflow.com/a/39741877/873282). So we know, the token is on your side - and not on ours to check review comments. I am aware that the PR title then changes often - however, it is more easy to manage (at least for me).

koppor commented:
@Ka0o0 While resolving the conflicts, I saw that a new string "Database mode" was inserted to the language file. We decided to use "Library" instead of "Database". See https://github.com/JabRef/jabref/issues/2095 (refs https://github.com/JabRef/jabref/pull/2748).

We use `%0 mode`, `New %0 library`. We have `Library properties`, so it is IMHO OK to add `Library mode`. Could you please change that @Ka0o0? This is IMHO the last thing TODO here, right @tobiasdiez?

Ka0o0 commented:
@koppor @tobiasdiez sorry for my late reply. I was on holiday this we. 
I have changed the strings now. Does this PR need anything else?

koppor commented:
@Ka0o0 Thank you for following up 🎉

tobiasdiez commented:
Thanks for your work on this issue, and I'm sorry that you came under friendly crossfire on how to best implement this feature which made the whole progress a bit longer than it needed be. I promise the next one will be more straightforward ;-)
